File: kotlin-typescript/karakum/nameResolvers/resolveInterfaceMethodTypeParameterConstraintName.js
import ts from "typescript";
import * as karakum from "karakum";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const typeParameter = getParent(node)
    if (!typeParameter) return null
    if (!ts.isTypeParameterDeclaration(typeParameter)) return null
    if (typeParameter.constraint !== node) return null

    const typeParameterName = typeParameter.name.text

    const method = getParent(typeParameter)
    if (!method) return null
    if (!ts.isMethodSignature(method)) return null
    if (!ts.isIdentifier(method.name)) return null

    const methodName = method.name.text

    const interfaceNode = getParent(method)
    if (!interfaceNode) return null
    if (!ts.isInterfaceDeclaration(interfaceNode)) return null

    const parentName = interfaceNode.name.text

    return `${karakum.capitalize(parentName)}${karakum.capitalize(methodName)}${karakum.capitalize(typeParameterName)}Constraint`
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/wrappersgenerator/typescript/nameResolvers/resolveInterfaceMethodTypeParameterConstraintName.kt
package wrappersgenerator.typescript.nameResolvers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.*

fun resolveInterfaceMethodTypeParameterConstraintName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))

    val typeParameter = ensureNotNull(typeScriptService.getParent(node))
    ensure(isTypeParameterDeclaration(typeParameter))
    ensure(typeParameter.constraint === node)
    val typeParameterName = typeParameter.name.text

    val method = ensureNotNull(typeScriptService.getParent(typeParameter))
    ensure(isMethodSignature(method))

    val methodNameNode = method.name
    ensure(isIdentifier(methodNameNode))
    val methodName = methodNameNode.text

    val interfaceNode = ensureNotNull(typeScriptService.getParent(method))
    ensure(isInterfaceDeclaration(interfaceNode))
    val parentName = interfaceNode.name.text

    "${parentName.replaceFirstChar { it.titlecase() }}${methodName.replaceFirstChar { it.titlecase() }}${typeParameterName.replaceFirstChar { it.titlecase() }}Constraint"
}
